src/Weeder/Main.hs

```diff
@@ -166,10 +271,19 @@ mainWithConfig hieExt hieDirectories requireHsFiles weederConfig@Config{ rootPat
       let hsFileExists = any ( hie_hs_file hieFileResult `isSuffixOf` ) hsFilePaths
        in requireHsFiles ==> hsFileExists
 
-  analysis <-
-    execStateT ( analyseHieFiles weederConfig hieFileResults' ) emptyAnalysis
+  pure hieFileResults'
+
 
+-- | Run Weeder on the given .hie files with the given 'Config'.
+--
+-- Returns a list of 'Weed's that can be displayed using
+-- 'formatWeed', and the final 'Analysis'.
+runWeeder :: Config -> [HieFile] -> ([Weed], Analysis)
+runWeeder weederConfig@Config{ rootPatterns, typeClassRoots, rootClasses, rootInstances } hieFiles =
```

Comment:
I also think this should go in `Weeder`, maybe as `analyzeHieFiles` (though I'd be OK with `runWeeder` too).

Reply:
Likewise with `runWeeder`, since it mostly interacts with the regular expressions provided by the user to decide what should be marked as a root, and only reads from some fields of `Analysis` instead of directly modifying it.

We could alternatively move the current `Weeder` module to something like `Weeder.Analysis`, and then move `runWeeder` to `Weeder.hs` - that way it's separate from both the `main` functions doing IO and the fine details of building the dependency graph.